src/transformers/generation_utils.py

```diff
@@ -1309,15 +1332,32 @@ def greedy_search(
                 outputs, model_kwargs, is_encoder_decoder=self.config.is_encoder_decoder
             )
 
+            # increase cur_len
+            cur_len = cur_len + 1
+
             # stop when there is a </s> in each sentence, or if we exceed the maximum length
             if unfinished_sequences.max() == 0:
+                print(f"{dist.get_rank()} unfinished_sequences")
```

Comment:
@stas00 - could you maybe try to do the following changes:

```
if unfinished_sequences.max() == 0 and not synced_gpus:
   break
```

instead of running: 

```
            for _ in range(max_length-cur_len):
                #_ = deepspeed(
                _ = self(
                    **model_inputs,
                    return_dict=True,
                    output_attentions=output_attentions,
                    output_hidden_states=output_hidden_states,
                )
```

I think this should work as well no? Since all appended tokens will then be padding tokens, they should be removed with `tokenizer.batch_decode(skip_special_tokens=True)` when doing `--predict_with_generate`

Reply:
There is one more break right after it with `stopping_criteria` so that one will need to do the same

The reason I coded this as an 2nd loop is to:

1. minimize the overhead - why calculate many things if we aren't using them, so to run just the bare minimum of code. There are 5 of these, and some of them run a lot more code than greedy_search.
2. to avoid complicating the existing code by separating the special needs into its own separate if branch so it's very easy to tell when studying the coder or running a debugger that this is a separate things I don't need to worry about
3. we must ensure that all forward calls will be repeated, I just wasn't sure if some model decides to take a shortcut if we feed it an input that it deems unnecessary to run forward with, so as a caution I was replaying that exact last input input that produced a full forward run. It's possible that no model has the smarts to do any internal shortcutting and my worry was unnecessary.

If you feel that solving this within the existing while loop is a better solution I will adapt to it.

Reply:
One more logic point to add. Since all gpus may finish their prediction sequence generation before `max_length` we can still finish early and to do that require something like:
```
while sync.item() > 0.0:
    p = model.forward(fake_input if am_i_done() else real_input)
    sync =torch.tensor(0.0 if am_i_done() else 1.0)
    torch.distributed.allreduce(sync)
```

which will communicate with all gpus to find the earlier point when all gpus may complete the loop.

Which means that this needs to be done as part of the main loop as you suggested and integrate the `break`s into this extra logic.

I will work on a new version.